satellite/gc/service.go

```diff
@@ -0,0 +1,147 @@
+	Enabled  bool          `help:"set if garbage collection is enabled or not" releaseDefault:"false" devDefault:"true"`
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int     `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a garbage collection bloom filter" releaseDefault:"0.1" devDefault:"0.1"`
+	ConcurrentSends   int     `help:"the number of nodes to concurrently send garbage collection bloom filters to" releaseDefault:"1" devDefault:"1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log    *zap.Logger
+	config Config
+	Loop   sync2.Cycle
+
+	transport    transport.Client
+	overlay      overlay.DB
+	metainfoLoop *metainfo.Loop
+}
+
+// RetainInfo contains info needed for a storage node to retain important data and delete garbage data
+type RetainInfo struct {
+	Filter       *bloomfilter.Filter
+	CreationDate time.Time
+	Count        int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, config Config, transport transport.Client, overlay overlay.DB, loop *metainfo.Loop) *Service {
+	return &Service{
+		log:    log,
+		config: config,
+		Loop:   *sync2.NewCycle(config.Interval),
+
+		transport:    transport,
+		overlay:      overlay,
+		metainfoLoop: loop,
+	}
+}
+
+// Run starts the gc loop service
+func (service *Service) Run(ctx context.Context) (err error) {
+
+	if !service.config.Enabled {
+		return nil
+	}
+
+	// TODO retrieve piece counts from overlay (when there is a column for them)
+	lastPieceCounts := make(map[storj.NodeID]int)
+
+	return service.Loop.Run(ctx, func(ctx context.Context) (err error) {
+		defer mon.Task()(&ctx)(&err)
+
+		pieceTracker := NewPieceTracker(service.log.Named("gc observer"), service.config, lastPieceCounts)
+
+		// collect things to retain
+		err = service.metainfoLoop.Join(ctx, pieceTracker)
+		if err != nil {
+			service.log.Error("error joining metainfoloop", zap.Error(Error.Wrap(err)))
+			return nil
+		}
+
+		// save piece counts for next iteration
+		for id := range lastPieceCounts {
+			delete(lastPieceCounts, id)
+		}
+		for id, info := range pieceTracker.retainInfos {
+			lastPieceCounts[id] = info.Count
+		}
+
+		// monitor information
+		for _, info := range pieceTracker.retainInfos {
+			mon.IntVal("node_piece_count").Observe(int64(info.Count))
+			mon.IntVal("retain_filter_size_bytes").Observe(info.Filter.Size())
+		}
+
+		// send retain requests
+		limiter := sync2.NewLimiter(service.config.ConcurrentSends)
+		for id, info := range pieceTracker.retainInfos {
+			id, info := id, info
+			limiter.Go(ctx, func() {
+				err := service.sendRetainRequest(ctx, id, info)
+				if err != nil {
+					service.log.Error("error sending retain info to node", zap.String("node ID", id.String()), zap.Error(err))
```

Comment:
Use `zap.Stringer("Node ID", id)` instead of `zap.String("node ID", id.String())`.

This is better for performance. `zap.Stringer` will call `id`'s `String()` method only if this message is going to be logged, i.e. log severity is at least set to Error.